cpp/ql/src/Security/CWE/CWE-295/SSLResultNotChecked.ql

```diff
@@ -0,0 +1,122 @@
+/**
+ * @name Certificate not checked
+ * @description Always check the result of certificate verification after fetching an SSL certificate.
+ * @kind problem
+ * @problem.severity error
+ * @security-severity 7.5
+ * @precision medium
+ * @id cpp/certificate-not-checked
+ * @tags security
+ *       external/cwe/cwe-295
+ */
+
+import cpp
+import semmle.code.cpp.valuenumbering.GlobalValueNumbering
+import semmle.code.cpp.controlflow.IRGuards
+
+/**
+ * A call to `SSL_get_peer_certificate`.
+ */
+class SSLGetPeerCertificateCall extends FunctionCall {
+  SSLGetPeerCertificateCall() {
+    getTarget().getName() = "SSL_get_peer_certificate" // SSL_get_peer_certificate(ssl)
+  }
+
+  Expr getSSLArgument() { result = getArgument(0) }
+}
+
+/**
+ * A call to `SSL_get_verify_result`.
+ */
+class SSLGetVerifyResultCall extends FunctionCall {
+  SSLGetVerifyResultCall() {
+    getTarget().getName() = "SSL_get_verify_result" // SSL_get_peer_certificate(ssl)
+  }
+
+  Expr getSSLArgument() { result = getArgument(0) }
+}
+
+/**
+ * Holds if the SSL object passed into `SSL_get_peer_certificate` is checked with
+ * `SSL_get_verify_result` entering `node`.
+ */
+predicate resultIsChecked(SSLGetPeerCertificateCall getCertCall, ControlFlowNode node) {
+  exists(Expr ssl, SSLGetVerifyResultCall check |
+    ssl = globalValueNumber(getCertCall.getSSLArgument()).getAnExpr() and
+    ssl = check.getSSLArgument() and
+    node = check
+  )
+}
+
+/**
+ * Holds if the certificate returned by `SSL_get_peer_certificate` is found to be
+ * `0` on the edge `node1` to `node2`.
+ */
+predicate certIsZero(
+  SSLGetPeerCertificateCall getCertCall, ControlFlowNode node1, ControlFlowNode node2
+) {
+  exists(GuardCondition guard, Expr cert |
+    cert = globalValueNumber(getCertCall).getAnExpr() and
+    (
+      exists(Expr zero |
+        zero.getValue().toInt() = 0 and
+        node1 = guard and
+        (
+          // if (cert == zero) {
+          guard.comparesEq(cert, zero, 0, true, true) and
+          node2 = guard.getATrueSuccessor()
+          or
+          // if (cert != zero) { }
+          guard.comparesEq(cert, zero, 0, false, true) and
+          node2 = guard.getAFalseSuccessor()
+        )
+      )
+      or
+      // if (cert) { }
+      guard = cert and
+      node1 = guard and
+      node2 = guard.getAFalseSuccessor()
+      or
+      // if (!cert) {
+      node1 = guard.getParent() and
+      node2 = guard.getParent().(NotExpr).getATrueSuccessor()
+    )
+  )
+}
+
+/**
+ * Holds if the SSL object passed into `SSL_get_peer_certificate` has not been checked with
+ * `SSL_get_verify_result` at `node`. Note that this is only computed at the call to
+ * `SSL_get_peer_certificate` and at the start and end of `BasicBlock`s.
+ */
+predicate certNotChecked(SSLGetPeerCertificateCall getCertCall, ControlFlowNode node) {
```

Comment:
Intead of rolling our own control-flow analysis here, could we instead use `StackVariableReachability`? Then `isSource` could be the call to `SSL_get_peer_certificate`, `isSink` could be the exit of the function, and `isBarrier` could be a call to `SSL_get_verify_result`.

Would that work?

It may be difficult to fit the `certIsZero` barrier into `SSL_get_peer_certificate`, but I think it's doable by having one `StackVariableReachability` for each call to `SSL_get_peer_certificate`.

Reply:
I tried a couple of libraries that didn't quite fit the problem, I may have overlooked `StackVariableReachability`.. Trying it now...

Reply:
> It may be difficult to fit the `certIsZero` barrier into `SSL_get_peer_certificate`, but I think it's doable by having one `StackVariableReachability` for each call to `SSL_get_peer_certificate`.

This turns out not to be a major issue, the existing GVN logic finds appropriate guards on certificates returned by `SSL_get_peer_certificate`, and I've never seen more than one certificate being used at once (in which case we'd have to worry about *which* certificate it is).

What is an issue is that the `certIsZero` barrier needs to be a barrier edge, not a barrier node.  `StackVariableReachability` does not appear to support barrier edges and I haven't found a sufficient workaround for this yet.  I suppose I could add it as a new feature, but that will then require a lot of testing.

Reply:
Fair enough. I appreciate that you looked into it! Let's postpone converting the logic to `StackVariableReachability` to a later point, then.